src/components/DynamicHelp/ProfilePageUsernameHelp/ProfilePageUsernameHelp.tsx

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) 2012-2022  Online-Go.com
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import * as React from "react";
+
+import * as dynamic_help from "dynamic_help_config";
+
+import { _ } from "translate";
+
+const HELP_SET = "guest-arrival-help-set";
+const ITEM = "profile-page-username-help";
+
+export function ProfilePageUsernameHelp(): JSX.Element {
```

Comment:
This is what I was referring to here:

> "each help item will need some individual layout customisation. It's not "cookie cutter popups", rather it is "bespoke help for each situation, based on a pattern".

As it turned out, both those components happened to look best positioned under the element that the fa-arrow-up is pointing at.   This does mean that in this specific "use case", really the only difference between the two classes should be the `ITEM` name, and the corresponding slightly different CSS for the slightly different positioning arising out of different anchors/insertion points:

![Screen Shot 2022-06-15 at 3 21 45 pm (2)](https://user-images.githubusercontent.com/61894/173753709-67c8dbd7-5a01-400b-b3a5-d27c3db32da1.png)

![Screen Shot 2022-06-15 at 3 25 57 pm (2)](https://user-images.githubusercontent.com/61894/173753727-44c30ca8-ac96-426c-9705-b66a3bd4cddd.png)


However _logically_ these are two completely separate help items in the set.  They are two separate instructions about how to get to where you change your username.   If that flow changes in its layout or mechanism, then these classes will diverge more in what they have to do - it's a "coincidence" that "the same code" works for these two cases.

The fact that "help items" in general are so similar is a benefit - I found that with the framework in place, I can wack them here and there very easily with copy and paste from the most similar other one, then tweaking.

There definitely ought to be room for improving reuse (such as visibility control), but my suspicion is that fully templated popups are a tantalising thought, but will come with different headaches ...  how to make the layout nice.

Reply:
I agree that it's a plus that Dynamic help supports bespoke widgets.  It makes a lot of sense for things like the email banner.  But when I look at these components, it seems like there are more reasons for them to stay the same than there are for them to diverge...

<img width="331" alt="Screen Shot 2022-06-14 at 11 37 55 PM" src="https://user-images.githubusercontent.com/25233703/173761361-bb57528f-7fe0-4f52-a01d-e7bacad96e77.png">
<img width="293" alt="Screen Shot 2022-06-14 at 11 38 25 PM" src="https://user-images.githubusercontent.com/25233703/173761376-9be2a56a-872f-468f-9fdf-0b927c8ab147.png">
<img width="285" alt="Screen Shot 2022-06-14 at 11 38 33 PM" src="https://user-images.githubusercontent.com/25233703/173761395-0fd9202b-4394-46e6-aa1b-87335d1a80c8.png">
<img width="312" alt="Screen Shot 2022-06-14 at 11 38 40 PM" src="https://user-images.githubusercontent.com/25233703/173761464-3427c229-df93-4d15-9ba0-d23e44aac4b5.png">
<img width="300" alt="Screen Shot 2022-06-14 at 11 38 47 PM" src="https://user-images.githubusercontent.com/25233703/173761490-8a148de9-0056-4889-9383-b4f46239271e.png">
<img width="314" alt="Screen Shot 2022-06-14 at 11 38 54 PM" src="https://user-images.githubusercontent.com/25233703/173761499-cb29e75f-c3af-43c6-8a9f-cfb5ac0a2d56.png">

Is it really worth the maintenance burden to have 6 of *almost* the same file? (18 if you count *.styl and index.ts)

Reply:
I guess this is a key feature of this design to debate.

Personally I suspect it is worth it, because the placement of each of those is carefully designed (in fact, you can see that the one pointing at Profile is not quite right yet, it falls off the right hand side).

Also, I think that the similarity we are seeing here is somewhat an artifact of the chosen flows.   When you have two or three clicks to do something (like "change your username") then you have to say almost the same thing three times, but in rather different places each time.

One possible way of looking at it is to say "lets see how we can commonise when we have more examples of what they will be like".

If we want a more templated approach, I think we'd need to come up with a less "arrow driven" way of pointing at things, because otherwise the combinations of arrow placement and direction relative to the text can get out of hand to parameterise.

Reply:
I think there's actually a more important design question.

The issue of "bespoke widgets propagates similar code" is second order.

The design issue to consider is besoke widgets vs uniform experience.

I chose bespoke widget _over_ uniform experience because I couldn't come up with a uniform experience scheme that would work reliably in OGS layout.

In some phone apps (for example accu-battery) they have rectangular help panesthat are like banners that insert themselves between normal app panes like flex elements in a flex: column container.   These are uniform - they always appear like that and take up that much space.    

I felt like I'd have to give up too much "ability to be specific about what I want to say" if we tried to do that.   You have to kind of "verbally indicate" what you are pointing at - and accu-battery suffers from this in the comprehensibility of its help messages.

"To change your username click here ->" is pretty appealing...

Reply:
Ironically, having made the case for what I did, my mind has started working on the opposing case, coming up with solutions for templating it, yet being able to position accurately and nicely.

The good thing about these initial ideas is that they seem to be compatible with the fully bespoke widgets.

One can stick with `<DynamicHelp.FullyBespokeOldHelpItem/>`  while also supporting 

```
<DynamicHelp.ParameterisedHelpItem
   set={'shiny-new-help-set'} 
   item={'another-almost-like-the-others'}
   arrow={'right'}
   text={'last click in this flow'} />
   ```
  
... so there's a backward compatible roadmap to parameterised joy :)

Reply:
> Personally I suspect it is worth it, because the placement of each of those is carefully designed (in fact, you can see that the one pointing at Profile is not quite right yet, it falls off the right hand side).

I worry a bit about anything that's "carefully designed" per say, as carefully designed things tend to break as folks come along and move stuff around and the site evolves. 

My first impression is that it feels a bit heavy to add new help. There's lots of boiler plate code that needs to be copied and touched in several files to add a new help bubble, so because of that I worry that it might not be used or maintained as much as we would like. That being said, I can see why you went down the path you did, for some aspects, though I can't help but feel like we can simplify things to make it easier to add new dialogs.

Something we should probably have in the bubbles is a "hide tutorial" link/button/thing that just turns them off - tutorial things like this are necessary for some and hated by others, so we just want to make it easy for folks to not have to go through it to get rid of them all if they don't want them.

I think we should call these out more in color, set the background to something a bit more colorful kind of thing too, and maybe add a bit of padding.

I think we should get rid of the "x" maybe? Just stick with click to dismiss, including perhaps clicking of the element we are drawing attention to (though that might be tricky and not worth it).

An alternative to the arrow would be to highlight whatever it is we're pointing at. One way do that would be to either pass a ref to the element we're pointing at and add a css class to it, and/or do that to children if we wrap the child in a help tag kind of a thing. Knowing that element could also be used for positioning too if we wanted to do that automatically, that way things just kinda move to the right place as the ui evolves.

Reply:
> Just stick with click to dismiss, including perhaps clicking of the element we are drawing attention to (though that might be tricky and not worth it).

Just a quick note: that clicking the element we're pointing at is designed to turn off the thing that was pointing at it (though this has to be coded for in each case).

Reply:
Another note:

> I think we should get rid of the "x" maybe?

I envisage wanting to be able to say "yeah yeah I know about this one, but I don't want to do it now".  That's what the 'x' is intended for.    Curiously, I'm not convinced it works properly all the time - although I have `stopPropagation` on that thing, sometimes I swear that I click the 'x' and get taken down the underlying (parent) link anyhow.

But ... the _intention_ is that folk should be able to dismiss the bubble without leaving the current page.